refactor/chain-index-methods

nachog00 opened the pull request:
## Motivation

The NonFinalizedState's `initialize()` and `sync()` methods had grown to 200+ and 179 lines respectively, with complex inline logic that was difficult to test and maintain. This refactoring breaks down these monolithic methods into focused, testable components.

## Solution

- **Modularized initialize()**: Extracted helper methods for listener setup, initial block resolution, and genesis block fetching
- **Simplified sync()**: Broke down sync logic into dedicated methods for staging, reorg handling, and event processing
- **Improved IndexedBlock creation**: Added unified helper methods and cleaner intermediate types
- **Enhanced interfaces**: Replaced tuple-based TryFrom with BlockWithMetadata for better type safety

### Tests

Existing integration tests continue to pass, verifying that the refactoring preserves functionality while improving code organization.

nachog00 committed:
Refactor NonFinalizedState::initialize method for better modularity

- Extract channel setup into StagingChannel helper struct
- Add BestTip::from_block() and NonfinalizedBlockCacheSnapshot::from_initial_block() helper methods
- Extract genesis block fetching into fetch_genesis_block() method
- Extract initial block resolution into resolve_initial_block() method
- Extract listener setup into setup_listener() method
- Reduce initialize() method from ~200 lines to ~40 lines of orchestration code

This improves code maintainability, testability, and readability by separating concerns
and eliminating the large inline genesis block processing logic.

nachog00 committed:
Improve initialize method by removing temp_self pattern

- Convert instance methods to static methods that take required parameters
- Remove awkward temp_self pattern that created incomplete instances
- Helper methods now take source and network parameters directly
- Cleaner, more explicit code that doesn't rely on incomplete self instances

The initialize method is now 27 lines of clear orchestration code.

nachog00 committed:
Eliminate duplicate block conversion logic and improve method clarity

- Extract block commitment conversion into BlockData::commitment_to_bytes() helper
- Refactor fetch_genesis_block to use existing IndexedBlock::try_from() instead of duplicating conversion logic
- Rename fetch_genesis_block to get_genesis_indexed_block for better clarity about what it returns
- Reduce method from ~165 lines to ~45 lines by reusing existing conversion infrastructure
- Clean up unused imports in non_finalised_state.rs

This eliminates substantial code duplication and ensures consistency in block conversion logic across the codebase.

nachog00 committed:
Refactor NonFinalizedState::sync method for better modularity

- Reduced main sync method from 179 lines to 22 lines
- Extracted fetch_main_chain_blocks() for block fetching and reorg detection
- Extracted handle_reorg() for dedicated reorg handling logic
- Extracted stage_new_blocks() for staging block operations
- Extracted handle_nfs_change_listener() for listener event processing
- Extracted process_nonbest_blocks() for non-best chain block handling
- Each method now has a single clear responsibility
- Improved code readability and maintainability

nachog00 committed:
Refactor IndexedBlock TryFrom with better intermediate types

- Add BlockMetadata and BlockWithMetadata intermediate types
- Extract complex transaction processing into focused helper methods
- Replace 150+ line monolithic TryFrom with modular approach
- Separate concerns: block data, transactions, indexing, commitment trees
- Mark old tuple-based interface as deprecated
- Maintain backward compatibility while improving maintainability
- All extraction methods are now separately testable

nachog00 committed:
Replace deprecated tuple TryFrom with BlockWithMetadata interface

- Remove old tuple-based TryFrom implementation for IndexedBlock
- Update all production code to use new BlockWithMetadata interface
- Add necessary imports for BlockMetadata and BlockWithMetadata
- Maintain the dangerous `as u32` casts for now (to be fixed by TreeSize PR)
- All call sites now use the cleaner, more maintainable interface

nachog00 committed:
Add unified helper methods to NonFinalizedState for IndexedBlock creation

- Add get_tree_roots_from_source() async method for fetching tree roots
- Add create_indexed_block_with_optional_roots() for genesis/sync cases
- Add create_indexed_block_with_required_roots() for finalized state cases
- Update block_to_chainblock() and get_genesis_indexed_block() to use new helpers
- Eliminates duplicate IndexedBlock creation patterns across the codebase
- Makes helper functions testable without requiring source setup
- Uses new TreeRootData type for cleaner interfaces

nachog00 committed:
Update imports and exports for TreeRootData across codebase

- Add TreeRootData to public exports in lib.rs
- Update linter-formatted code in migrations.rs and non_finalised_state.rs

idky137 commented on zaino-state/src/chain_index/types.rs:
```diff
@@ -1362,195 +1377,305 @@ impl
+#[derive(Debug, Clone)]
+pub struct TreeRootData {
+    /// Sapling tree root and size
+    pub sapling: Option<(zebra_chain::sapling::tree::Root, u64)>,
```
Hey I think we have been very clear that this pattern is not allowed in this module. Types must have thier own internals and use from / into for zebra or librustzcash types.

nachog00 replied:
This just bundles data into named types to pass onto `TryFrom<zebra's_block,...> for IndexedBlock`. It isn't meant to store data in Db nor does it alter any of the structs that read/write to database.

If you pay attention, you'll see it's no different from what was there before with the nameless tuple interface for the conversion.

Maybe you should reassess your rules ass it doesn't make much sense that they are set for the module but coherently for types that read/write to db.

AloeareV replied:
This rule makes sense if (and only if) `chain_index/types.rs` is purely for database types (which it currently is). If it is, where do we put chain_index types that aren't database types? We could rename `types` to `database_types`, or make a new `non_database_types` or something along those lines...

idky137 commented on zaino-state/src/chain_index/types.rs:
```diff
@@ -1362,195 +1377,305 @@ impl
+#[derive(Debug, Clone)]
+pub struct BlockMetadata {
+    /// Sapling commitment tree root
+    pub sapling_root: zebra_chain::sapling::tree::Root,
```
similar to above

idky137 commented on zaino-state/src/chain_index/types.rs:
```diff
@@ -1362,195 +1377,305 @@ impl
+    /// Sapling tree size
+    pub sapling_size: u32,
+    /// Orchard commitment tree root
+    pub orchard_root: zebra_chain::orchard::tree::Root,
```
similar to above

idky137 commented on zaino-state/src/chain_index/types.rs:
```diff
@@ -1362,195 +1377,305 @@ impl
+    /// Parent block's chainwork
+    pub parent_chainwork: ChainWork,
+    /// Network for block validation
+    pub network: zebra_chain::parameters::Network,
```
similar to above

idky137 reviewed (dismissed):
Hey this looks like it is breaking rules we have set out for the types module.

idky137 commented:
After looking this looks like its making a lot of changes to the types.rs module without following the clear rules set out.

nachog00 commented:
> Hey looks like interesting work!
> 
> Is this related to [Zaino, The Zallet Release](https://github.com/orgs/zingolabs/projects/19) at all or something new?

It is related. The aim of this work is to enable more atomicity in the code, in order to enable currently impossible unit testing, which in turn is needed to showcase current unknown/silent bug classes.

nachog00 committed:
Merge branch 'dev' into feature/refactor-initialize-method

AloeareV commented:
> well, there's a PR that merged not long ago that starts the migration from the single massive types.rs module onto smaller ones. We could simply have types::db and types::anything else.

I'd be on-board with this

nachog00 committed:
Refactor block metadata handling

- Updated `v0.rs` to use `BlockMetadata` and `BlockWithMetadata` for clearer block metadata handling
- Refactored `v1.rs` to implement `BlockMetadata` creation, separating metadata from block conversion
- Streamlined the process of creating `IndexedBlock` from metadata, improving code readability and maintainability
- Changes ensure better encapsulation and potential for easier future modifications in block processing logic

nachog00 committed:
Refactor and cleanup finalised_state test v1.rs
- Reorganize imports for better readability
- Remove unnecessary parentheses when creating metadata tuples
- Introduce `chain_block` initialization for improved code clarity

nachog00 committed:
Refactor tests to use BlockWithMetadata
- Updated 'migrations.rs', 'v0.rs', and 'v1.rs' to use `BlockWithMetadata` class
- Replaced inline metadata creation with dedicated BlockMetadata instances
- Improved readability and maintainability of test functions

The changes help in streamlining the creation and manipulation of block metadata, ensuring consistency across tests.

nachog00 committed:
Merge branch 'feature/refactor-initialize-method' of github.com:zingolabs/zaino into feature/refactor-initialize-method

nachog00 committed:
Fix clippy warnings

- Remove unnecessary .clone() calls on ChainWork (implements Copy trait)
- Replace useless format!() with .to_string()
- Remove needless borrow in test function

nachog00 committed:
Merge branch 'dev' into feature/refactor-initialize-method

nachog00 committed:
Initialize types::db module for database-serializable types

- Create types/db.rs module for types implementing ZainoVersionedSerialise
- Add module documentation outlining strict versioning rules
- Update types.rs to include db module

nachog00 committed:
Move commitment module into types::db

- Relocate commitment.rs to types/db/commitment.rs
- Update db.rs to declare and re-export commitment types
- Fix import paths in commitment module
- Update types.rs to re-export from db module

nachog00 committed:
Create empty db submodules for organizing database types

- Add db/block.rs for block-related types
- Add db/transaction.rs for transaction types
- Add db/address.rs for address and UTXO types
- Add db/shielded.rs for Sapling/Orchard types
- Add db/primitives.rs for basic primitives
- Update db.rs to declare all new submodules

nachog00 committed:
Copy types.rs to types/db/legacy.rs for refactoring

Create db/legacy.rs as a copy of types.rs containing all database-
serializable types. This bloated file will be incrementally refactored
by extracting types into focused modules (block.rs, transaction.rs, etc).

Re-export all legacy types from db module for now to maintain compatibility.

nachog00 committed:
Remove non-DB types from db/legacy.rs

Remove types that don't implement ZainoVersionedSerialise from db/legacy.rs:
- BestChainLocation
- NonBestChainLocation
- TreeRootData
- BlockMetadata
- BlockWithMetadata

These types remain in the original types.rs as they are helper/wrapper
types not intended for database serialization.

nachog00 committed:
Fix db/legacy.rs module structure and imports

- Remove invalid submodule declarations (db, primitives)
- Consolidate imports from crate::chain_index::encoding
- Import CommitmentTree types from super::commitment
- Remove duplicate derive attributes on CompactTxData

The legacy.rs file now compiles successfully and contains only
database-serializable types (those implementing ZainoVersionedSerialise).

nachog00 committed:
Refactor types.rs into module index with helpers and db modules

- Create types/helpers.rs for non-database helper types:
  - BestChainLocation, NonBestChainLocation
  - TreeRootData, BlockMetadata, BlockWithMetadata

- Transform types.rs into clean module index file:
  - Re-exports db types from db/legacy
  - Re-exports helper types from helpers module
  - Clear documentation of module organization

- Fix non_finalised_state.rs to use height() method instead of field access

- Add Clone derive and doc comment to TreeRootData

Field visibility unchanged - pub(super) fields accessible via re-exports
from parent types module. types.rs is now a proper module index file.

nachog00 committed:
Add missing ZainoVersionedSerialise impl for IndexedBlock

The ZainoVersionedSerialise implementation for IndexedBlock was
accidentally omitted when extracting helper types from types.rs.
This impl belongs in db/legacy.rs since IndexedBlock is a database-
serializable type.

Fixes test failures where from_bytes() and to_bytes() methods were
not found for IndexedBlock.

nachog00 committed:
Run cargo fmt

nachog00 committed:
Merge branch 'feature/refactor-initialize-method' of github.com:zingolabs/zaino into feature/refactor-initialize-method

nachog00 committed:
Merge branch 'dev' into feature/refactor-initialize-method

nachog00 committed:
Merge branch 'dev' into feature/refactor-initialize-method

nachog00 committed:
Merge branch 'dev' into feature/refactor-initialize-method

nachog00 committed:
Refactor v0_to_v1_full function in migrations.rs
- Renamed 'parent_chain_work' to 'parent_chainwork' for consistency
- Refactored 'IndexedBlock::try_from' to improve readability by using 'BlockMetadata::new'
- Streamlined variable naming: 'sapling_root_size' to 'sapling_size' and 'orchard_root_size' to 'orchard_size'
- Enhanced clarity by consolidating block metadata initialization and 'IndexedBlock' creation

Improves code readability and consistency across the project.

nachog00 committed:
Refactored block construction to include metadata explicitly

- Updated migrations.rs: Changed IndexedBlock construction to include a separate creation of BlockMetadata before combining with zebra_block
- Updated v1.rs: Applied similar refactoring by creating BlockMetadata before calling IndexedBlock::try_from, enhancing code clarity and separation of concerns

nachog00 committed:
Update documentation references

- Removed unused reference links in `types.rs`
- Updated module references in `db/legacy.rs` from `types::` to `super::`

nachog00 committed:
Merge branch 'dev' into feature/refactor-initialize-method

nachog00 committed:
Change function call in non_finalised_state.rs
- Rename function call from `create_indexed_block_with_required_roots` to `create_indexed_block_with_optional_roots`
- Reflects a change in requirement for block creation, allowing for optional roots instead

nachog00 committed:
Remove unused function in non_finalised_state.rs
- Deleted `create_indexed_block_with_required_roots` function
- Function was not being used and removed for code cleanliness

nachog00 committed:
Document activation height validation issue in create_indexed_block_with_optional_roots

Add TODO comment referencing issue #604. The current uniform unwrap_or_default()
approach masks potential bugs since Sapling and Orchard have different activation
heights (419,200 vs 1,687,104 on mainnet).

nachog00 committed:
Remove unused extract_required method from TreeRootData

This method had no valid use cases in the codebase. Finalized state does its own
manual validation with custom error messages, and non-finalized state should use
optional/default handling since tree roots can legitimately be None for
pre-activation blocks.
